libraries/lib-mixer/Mix.cpp

```diff
@@ -150,45 +363,76 @@ Mixer::Mixer(
       std::accumulate(
          mInputs.begin(), mInputs.end(), 0,
          [](auto sum, const auto& input) {
-            return sum + input.stages.size() * input.pSequence->NChannels();
+            return sum + input.stages.size();
          }) +
       nMasterStages;
    mSettings.reserve(nStages);
    mStageBuffers.reserve(nStages);
 
    size_t i = 0;
+   std::vector<std::unique_ptr<DownmixSource>> mixdownSources;
    for (auto &input : mInputs) {
       const auto &sequence = input.pSequence;
       if (!sequence) {
          assert(false);
          break;
       }
-      auto increment = finally([&]{ i += sequence->NChannels(); });
 
       auto &source = mSources.emplace_back(sequence, BufferSize(), outRate,
-         warpOptions, highQuality, mayThrow, mTimesAndSpeed,
-         (pMixerSpec ? &pMixerSpec->mMap[i] : nullptr));
+         warpOptions, highQuality, mayThrow, mTimesAndSpeed);
       AudioGraph::Source *pDownstream = &source;
       for (const auto &stage : input.stages)
          if (
             auto& pNewDownstream =
-               RegisterEffectStage(*pDownstream, stage, outRate))
+               RegisterEffectStage(*pDownstream, sequence->NChannels(), stage, outRate))
          {
             pDownstream = pNewDownstream.get();
          }
-      mDecoratedSources.emplace_back(Source{ source, *pDownstream });
+      mixdownSources.emplace_back(
```

Comment:
Looking at this I first wondered why a "sequence source" might be the last node of an track's realtime effect stack. Now I see that, in fact, it exposes the processed audio, but uses `*sequence` to get information about channels and gains.
I think it'd be less confusing if it were called `ProcessedSequenceDownmixSource` and it didn't take a `WideSampleSequence` reference but only the metadata it needs about it, to make it really clear that it's not going to pull data from that reference.

Reply:
Probably the confusion comes from the fact that `DownmixSource` isn't in same hierachy with `AudioSource`
`SequenceDownmixSource`. I'll add more comments that should clarify that and extract it into a sepatate source file as you suggested